.gitlab/e2e_test_junit_upload.yml

```diff
@@ -74,18 +80,46 @@ e2e_test_junit_upload:
     - new-e2e-agent-platform-install-script-upgrade7-ubuntu-x86_64
     - new-e2e-agent-platform-install-script-upgrade6-ubuntu-x86_64
     - new-e2e-agent-platform-install-script-upgrade7-ubuntu-iot-agent-x86_64
+    - new-e2e-windows-agent-msi-windows-server-a6-x86_64
+    - new-e2e-windows-agent-msi-windows-server-a7-x86_64
+
+e2e_pre_test_junit_upload:
+  extends: .e2e_test_junit_template
+  rules:
+  # using the same rules as the kitchen_tests_upload job
+    - !reference [.on_e2e_changes_or_manual]
+  when: always
+  stage: e2e_pre_test_junit_upload
+  dependencies:
+    # We need to exhaustively list all the `new-e2e-…` jobs that produce junit reports here
+    # to avoid downloading all the artifacts of all the jobs of all the previous stages.
+    - e2e_pre_test
+
+e2e_container_test_junit_upload:
+  extends: .e2e_test_junit_template
+  rules:
+    - !reference [.on_container_or_e2e_changes_or_manual]
+  when: always
+  stage: e2e_test_junit_upload
+  dependencies:
+    # We need to exhaustively list all the `new-e2e-…` jobs that produce junit reports here
+    # to avoid downloading all the artifacts of all the jobs of all the previous stages.
+    - new-e2e-containers
+
+e2e_test_junit_upload:
+  extends: .e2e_test_junit_template
+  rules: !reference [.on_main]
+  when: always
+  stage: e2e_test_junit_upload
+  dependencies:
+    # We need to exhaustively list all the `new-e2e-…` jobs that produce junit reports here
+    # to avoid downloading all the artifacts of all the jobs of all the previous stages.
+    - new-e2e-agent-shared-components-main
+    - new-e2e-agent-subcommands-main
+    - new-e2e-language-detection-main
```

Comment:
Moved to `e2e_container_test_junit_upload` that runs with the same rule as `new-e2e-containers`

```yaml
e2e_container_test_junit_upload:
  extends: .e2e_test_junit_template
  rules:
    - !reference [.on_container_or_e2e_changes_or_manual]
  when: always
  stage: e2e_test_junit_upload
  dependencies:
    # We need to exhaustively list all the `new-e2e-…` jobs that produce junit reports here
    # to avoid downloading all the artifacts of all the jobs of all the previous stages.
    - new-e2e-containers
```

`always` allows to run the job as soon as the dependency is done. This will upload test results from branch runs too

Reply:
gitlab rules are so cryptic !